### PR TITLE
Add missing residue validation to `ProteinComponent`

### PR DESCRIPTION
**Pull request by jthorton**
This PR adds a validate method to the `ProteinComponent` and derived classes which attempts to find missing residues and capping groups in the input via large inter-residue bond distances, which cause crashes downstream. 

## TODO
- make sure all Industry benchmark inputs pass validation
- do we need to account for box vectors in the derived solvated components? 

Tips
* Comment "pre-commit.ci autofix" to have pre-commit.ci atomically format your PR.
  Since this will create a commit, it is best to make this comment when you are finished with your work.

## LLM / AI generated code disclosure

LLMs or other AI-powered tools (beyond simple IDE use cases) were used in this contribution: yes / no
If yes, please provide details here: Yes, Codex was used to improve the bond-filtering code to try to only hit peptide bonds. 

## Checklist
* [x] Added a ``news`` entry
* [x] Filled in the AI-generated code disclosure.

## Developers certificate of origin
- [x] I certify that this contribution is covered by the MIT License [here](https://github.com/OpenFreeEnergy/openfe/blob/main/LICENSE) and the **Developer Certificate of Origin** at <https://developercertificate.org/>.

### CODE REVIEW

**Comment by jthorton** [comment]
pre-commit.ci autofix

File: src/gufe/tests/data/pxr_1nrl.pdb
Comment (jthorton) [inline src/gufe/tests/data/pxr_1nrl.pdb:1]: Should this be compressed?

Comment (IAlibay) [reply src/gufe/tests/data/pxr_1nrl.pdb:1]: Yeah always better - especially if we're one day going to do a pypi release, smaller data is such a life saver.

File: news/protien_validation.rst [hunk 1]
```diff
@@ -0,0 +1,23 @@
+**Added:**
+
+* ProteinComponent and derived classes (SolvatedPDBComponent and ProteinMembraneComponent) can attempt to find missing residues and capping groups in the input via large inter-residue bond distances via the ``validate`` method.
```
Comment (IAlibay) [inline news/protien_validation.rst:3]: Could you reword this to be more informative to a user - e.g. an error will be raised if you have missing residues

File: src/gufe/components/solvatedpdbcomponent.py [hunk 2]
```diff
@@ -483,5 +492,5 @@ def validate(
             raise ComponentValidationError(
                 "ProteinMembraneComponent validation failed:\n"
                 + "\n".join(f"- {e}" for e in errors)
-                + "\nThis usually indicates missing solvent or incorrect box vectors."
+                + "\nThis usually indicates missing solvent, incorrect box vectors or missing residues or capping groups."
```
Comment (IAlibay) [inline src/gufe/components/solvatedpdbcomponent.py:495]:
```suggestion
                + "\nThis usually indicates missing solvent, incorrect box vectors or an incorrect protein structure."
```
[nit] simplifying to avoid too many `or` entries.

File: news/protien_validation.rst
Comment (IAlibay) [inline news/protien_validation.rst:1]: typo in file name: protien -> protein

File: src/gufe/components/proteincomponent.py [hunk 3]
```diff
@@ -634,3 +637,80 @@ def _to_dict(self) -> dict:
...
+
+        possible_bad_bonds = []
+        for b in rd_mol.GetBonds():
+            a1, a2 = b.GetBeginAtom(), b.GetEndAtom()
+            m1, m2 = a1.GetMonomerInfo(), a2.GetMonomerInfo()
+            if m1 is None or m2 is None:
+                continue
+
+            # only inter-residue, same-chain bonds
+            if m1.GetResidueNumber() == m2.GetResidueNumber():
```
Comment (IAlibay) [inline src/gufe/components/proteincomponent.py:673]: This will silently skip insertion codes - you will need to assert on the output of `GetResidueNumber` and `GetInsertionCode`.

File: src/gufe/components/proteincomponent.py [hunk 4]
```diff
@@ -634,3 +637,80 @@ def _to_dict(self) -> dict:
...
+                f"{m1.GetChainId()}:{m1.GetResidueName()}{m1.GetResidueNumber()}:{m1.GetName().strip()} - "
+                f"{m2.GetChainId()}:{m2.GetResidueName()}{m2.GetResidueNumber()}:{m2.GetName().strip()} = {d:.2f} A"
+                for m1, m2, d in possible_bad_bonds
+            )
+            raise ComponentValidationError(
+                "Detected long inter-residue peptide C-N bonds, likely uncapped/missing residues. Check the following bonds:\n"
+                + msg
+            )
+
+    def validate(self, *, peptide_bond_cutoff: Quantity = 2.0 * offunit.angstrom):
```
Comment (IAlibay) [inline src/gufe/components/proteincomponent.py:702]: The a2a input is failing here because the C-N bond is 2.6 A instead of 2.0. I would suggest increasing this to 3.0 A, that's just over the usual distance between C & N in a C-CA-N backbone connection.

```suggestion
    def validate(self, *, peptide_bond_cutoff: Quantity = 3.0 * offunit.angstrom):
```

File: src/gufe/components/proteincomponent.py [hunk 5]
```diff
@@ -634,3 +637,80 @@ def _to_dict(self) -> dict:
...
+            # we want peptide C-N single bonds only
+            if not b.GetBondType() == Chem.BondType.SINGLE:
+                continue
+
+            names = {m1.GetName().strip(), m2.GetName().strip()}
+            if names != {"C", "N"}:
+                continue
+
+            i, j = a1.GetIdx(), a2.GetIdx()
+            d = np.linalg.norm(xyz[i] - xyz[j])
```
Comment (IAlibay) [inline src/gufe/components/proteincomponent.py:687]: We definitely need to account for PBC if we're going to do this on SolvatedPDBComponents.

I can't remember if we allow for triclinic cellls as inputs - if we do, this will be a bit of a pain to deal with, if not minimizing the vector isn't too bad.

File: src/gufe/components/solvatedpdbcomponent.py [hunk 6]
```diff
@@ -116,18 +117,23 @@ def validate(
...
+            * If the protein has missing residues without capping groups or model breaks.
         """
         if self.density < min_density:  # type: ignore
             raise ComponentValidationError(
                 "Estimated system density is very low.\n  "
                 f"Density: {self.density:.3f} (expected ≥ {min_density}). "
                 "This usually indicates missing solvent or incorrect box vectors."
             )
+        # run the bond distance validation
+        super().validate(peptide_bond_cutoff=peptide_bond_cutoff)
```
Comment (IAlibay) [inline src/gufe/components/solvatedpdbcomponent.py:136]: You'll probably need to wrap this around a try/except (and move it up) like we do for ProteinMembraneComponent, that way we don't immediately fail at the first validation.

File: src/gufe/components/proteincomponent.py [hunk 7]
```diff
@@ -634,3 +637,80 @@ def _to_dict(self) -> dict:
...
+            If any inter-residue peptide C-N bonds are found to be longer than the specified cutoff,
+            indicating likely uncapped or missing residues.
+        """
+        rd_mol = self._rdkit
+        conf = rd_mol.GetConformer(0)
+        xyz = np.asarray(conf.GetPositions())  # angstrom
+        bond_threshold = peptide_bond_cutoff.m_as(offunit.angstrom)
+
+        possible_bad_bonds = []
+        for b in rd_mol.GetBonds():
```
Comment (IAlibay) [inline src/gufe/components/proteincomponent.py:666]: Going bond first imparts a pretty big performance penalty with rdkit. I'll make a PR into this PR with a suggested atom-first approach.

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: src/gufe/components/proteincomponent.py [hunk 8]
```diff
@@ -634,3 +637,106 @@ def _to_dict(self) -> dict:
...
+            return
+
+        if box_vectors is None:
+            idx_i = np.array([c[2] for c in candidates])
+            idx_j = np.array([c[3] for c in candidates])
+            distances = np.linalg.norm(conf[idx_i] - conf[idx_j], axis=1)
+        else:
+            from openmm.app.internal import compiled
+            # Use OpenMM's compiled functions to compute distances with PBC
+            periodic_distance_func = compiled.periodicDistance(box_vectors.m_as(offunit.nanometer))
```
Comment (IAlibay) [inline src/gufe/components/proteincomponent.py:713]: This is great, didn't know OpenMM had this available. 

The only snag I can see here is that it's cementing our OpenMM dependency. For now it's fine, longer term we might need to consider vendoring / creating our own distance methods.

Comment (jthorton) [reply src/gufe/components/proteincomponent.py:713]: Yeah exactly I figured we could use or vendor this the same way we do with PDB, its probably best to use it till we need to vendor though?

File: src/gufe/components/solvatedpdbcomponent.py (hunk 6 above)
Comment (jthorton) [reply src/gufe/components/solvatedpdbcomponent.py]: Went with something a little different to make it easier to get the errors, hopefully it makes extending the validators easier.

**Comment by jthorton** [comment]
pre-commit.ci autofix

File: src/gufe/components/proteincomponent.py (hunk 4 above)
Comment (jthorton) [reply src/gufe/components/proteincomponent.py]: Yeah that cap is not great so I agree that this might be too strict but this did catch some cases where atoms in caps had been labelled the wrong way, which would have been missed at 3 A, I guess it easy to adjust if we want to run more strict validation, though. So lets go with 3 for now.

File: src/gufe/components/proteincomponent.py (hunk 8 above)
Comment (IAlibay) [reply src/gufe/components/proteincomponent.py:713]: Cython is a big pain to package - it means building wheels in a completely different way and has really annoying maintenance burdens. Hopefully we do not need to vendor it.

File: src/gufe/tests/test_solvatedpdbcomponent.py [hunk 9]
```diff
@@ -164,6 +164,18 @@ def test_validate_low_density_raises(self, instance):
         with pytest.raises(ComponentValidationError, match="density is very low"):
             instance.validate(min_density=500 * offunit.gram / offunit.liter)
 
+    def test_validate_low_density_and_missing_residues_raises(self, PDB_pxr_1nrl_path):
```
Comment (IAlibay) [inline src/gufe/tests/test_solvatedpdbcomponent.py:167]: Can you also add a test for a system without box vectors?

File: src/gufe/tests/data/pxr_1nrl.pdb.gz
Comment (IAlibay) [inline src/gufe/tests/data/pxr_1nrl.pdb.gz]: 185 KB is quite a lot for a gzipped PDB - are there waters we can throw away here?

File: src/gufe/components/solvatedpdbcomponent.py [hunk 10]
```diff
@@ -116,18 +152,21 @@ def validate(
...
-                "Estimated system density is very low.\n  "
-                f"Density: {self.density:.3f} (expected ≥ {min_density}). "
-                "This usually indicates missing solvent or incorrect box vectors."
-            )
+        errors = self._collect_validation_errors(
+            min_density=min_density,
+            peptide_bond_cutoff=peptide_bond_cutoff,
+        )
+        if errors:
+            raise ComponentValidationError("...\n" + "\n".join(f"- {e}" for e in errors))
```
Comment (IAlibay) [inline src/gufe/components/solvatedpdbcomponent.py]: Is the `...` intended here? Shoudl it have the component name like in the Membrane component?

File: src/gufe/components/solvatedpdbcomponent.py [hunk 11]
```diff
@@ -104,10 +106,44 @@ def density(self) -> Quantity:
 
         return total_mass.to("gram") / volume_L
 
+    def _collect_validation_errors(self, *, min_density: Quantity, peptide_bond_cutoff: Quantity) -> list[str]:
+        """
+        Helper function to collect validation errors without raising exceptions,
+        useful for derived classes that want to extend validation.
```
Comment (IAlibay) [inline src/gufe/components/solvatedpdbcomponent.py]:
```suggestion
        Helper function to collect validation errors without raising exceptions.
```
Since it's not a base class, I would remove this bit to avoid confusion later on.

File: src/gufe/components/solvatedpdbcomponent.py [hunk 12]
```diff
@@ -104,10 +106,44 @@ def density(self) -> Quantity:
...
+        peptide_bond_cutoff : openff.units.Quantity
+            The cutoff used to detect large peptide bond distances in the protein due to missing residues or capping groups.
+
+        Returns
+        -------
+        list[str]
+            List of validation error messages. Empty if no errors were found.
+        """
+        errors = []
+        if self.density.m_as(_g_per_ml) < min_density.m_as(_g_per_ml):
```
Comment (IAlibay) [inline src/gufe/components/solvatedpdbcomponent.py]: Given how little code is involved here, I don't necessarily understand why this isn't being done in `validate`. From a "future developer" standpoint, it seems to be adding more complexity than is necessary.

I.e. you're really just doing the validation here, it would be better to just do the validation in the validate method.

File: src/gufe/components/solvatedpdbcomponent.py [hunk 13]
```diff
@@ -445,11 +484,43 @@ def _count_waters(rdkit_mol: Mol) -> int:
         frags = Chem.rdmolops.GetMolFrags(rdkit_mol, asMols=False)
         return sum(ProteinMembraneComponent._is_water_fragment(rdkit_mol, frag) for frag in frags)
 
+    def _collect_validation_errors(
+        self, *, min_density: Quantity, peptide_bond_cutoff: Quantity, min_waters: int = 50
+    ) -> list[str]:
+        """
+        Helper function to collect validation errors without raising exceptions,
+        useful for derived classes that want to extend validation.
```
Comment (IAlibay) [inline src/gufe/components/solvatedpdbcomponent.py]:
```suggestion
        Helper function to collect validation errors without raising exceptions.
```
As above, although I would also be a proponent for now having this and just doing all of this in `validate`.

File: src/gufe/components/solvatedpdbcomponent.py [hunk 14]
```diff
@@ -445,11 +484,43 @@ def _count_waters(rdkit_mol: Mol) -> int:
...
+
+        Parameters
+        ----------
+        min_density : openff.units.Quantity
+            Minimum number of water molecules. Default: 50
```
Comment (IAlibay) [inline src/gufe/components/solvatedpdbcomponent.py]: Wrong description?

File: src/gufe/tests/test_proteincomponent.py [hunk 15]
```diff
@@ -418,9 +419,11 @@ def test_protein_valence_error(self):
             with mock.patch("rdkit.Chem.rdchem.PeriodicTable.GetValenceList", return_value=[10000000]):
                 self.cls.from_pdb_file(ALL_PDB_LOADERS["3tzr_rna"]())
 
-    def test_empty_validate(self, instance):
```
Comment (IAlibay) [inline src/gufe/tests/test_proteincomponent.py:421]: Any reason for removing this test?

File: src/gufe/tests/conftest.py [hunk 16]
```diff
@@ -401,3 +401,9 @@ def stereo_chem_mapping():
 def atropisomeric_sdf():
     with importlib.resources.path("gufe.tests.data", "atropisomeric_mol.sdf") as f:
         yield str(f)
+
+
+@pytest.fixture()
+def PDB_pxr_1nrl_path():
```
Comment (IAlibay) [inline src/gufe/tests/conftest.py:407]: This doesn't test the "bonds stretched over a periodic image", can we either have a test where we shift the cap across the periodic image, or another test where the system is broken across the periodic boundary?

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: src/gufe/components/solvatedpdbcomponent.py (hunk 10 above)
Comment (jthorton) [reply src/gufe/components/solvatedpdbcomponent.py]: Removed!

File: src/gufe/components/solvatedpdbcomponent.py (hunk 12 above)
Comment (jthorton) [reply src/gufe/components/solvatedpdbcomponent.py]: Fair enough this was to work around the formatting in the error message, moved it all to the validate methods and removed the helper functions.

File: src/gufe/components/solvatedpdbcomponent.py (hunk 14 above)
Comment (jthorton) [reply src/gufe/components/solvatedpdbcomponent.py]: removed!

File: src/gufe/tests/test_proteincomponent.py (hunk 15 above)
Comment (jthorton) [reply src/gufe/tests/test_proteincomponent.py:421]: The  ProteinComponent now has validation I can add it back to check it passes but its no longer an empty validator test.

File: src/gufe/tests/conftest.py (hunk 16 above)
Comment (jthorton) [reply src/gufe/tests/conftest.py:407]: Yeah this is more do we get the same answer, I'll add another test for the PBC.

**Comment by jthorton** [comment]
pre-commit.ci autofix

File: src/gufe/tests/test_proteincomponent.py (hunk 15 above)
Comment (jthorton) [reply src/gufe/tests/test_proteincomponent.py:421]: Add the test back to make sure a valid component passes.

File: src/gufe/tests/test_solvatedpdbcomponent.py (hunk 9 above)
Comment (jthorton) [reply src/gufe/tests/test_solvatedpdbcomponent.py:167]: Added!

**Comment by jthorton** [comment]
pre-commit.ci autofix

**Review by IAlibay — APPROVED** [review]
LGTM, thanks!
